### ISSUE TITLE
Improve / refactor styles system

### ISSUE THREAD

**Issue by pastk** [organicmaps/organicmaps#4317]
The current styles system based on Mapcss and kothic is hard to work with.
Its unpredictable, full of hacks, there are many duplicate and simply not working rules in the files, etc.

Requirements for a better system:
- predictable and easily debuggable, makes it hard to make mistakes and break things
- style files are nice, clean & compact
- promote stylistic standards, e.g. standard font and icon sizes for various zoom leves
- support easy "sub-styling", i.e. add new styles by specifying only changes over the base style
- simple enough to be used by enthusiastic users (think custom styles)
- powerful/flexible enough to describe the styles without much repetition
- suitable for a creation of a (mobile?) visual styles designer app in future

In this ticket first let's document the current system design and the biggest issues with it.
And then we can come with a better design perhaps.

**Comment by pastk** [comment]
A big issue is that mapcss files use OSM tags for rule selectors instead of internal OM types, e.g. it uses `[highway=footway][sac_scale=hiking]` instead of `highway-footway-hiking`, so one needs to consult `mapcss-mapping.csv` and `replaced_tags.txt` - it makes matching more error prone and harder to debug (as the compiled drule files use internal OM types). We should just use internal types in style files directly, it'll solve a lot of issues.

At the moment for each change in mapcss files we really have to double-check with text "printouts" of compiled drules files to ensure that changes are as expected and nothing unpredictable happen. And sometimes its hard to do, e.g. see OSM tags / OM types above or  different z-index / priority numbers(see #4314). Ideally we won't need that double-checking - is it possible? How to make styles compilation predictable?

The style files look too verbose, there are a lot of repetitive boilerplate declarations used e.g. for each zoom level of each feature. The syntax could be re-designed to be more compact for our use patterns.
We don't really need to keep compatibility with Mapcss spec/syntax and kothic parser implementation as they seem too powerful (and thus too complex) for our use case, i.e. we use only a subset of their features.

To promote standard things like font sizes we can use constants more extensively (just like we do with colors).

We don't have tools to validate and auto-format the files, hence they're quite messy (inconsistent formatting, duplicate rules, rules changing nothing, etc.)

**Comment by pastk** [comment]
Related:
- #4314 
- #2026

**Comment by pastk** [comment]
@vng @biodranik @endim8 @dvdmrtnz

**Comment by biodranik** [comment]
How is it solved in other map-related projects?

For humans it is way more easier to work with OpenStreetMap.org tags instead of our internal types. Can we somehow make it easier/automate/validate?

**Comment by pastk** [comment]
> For humans it is way more easier to work with OpenStreetMap.org tags instead of our internal types. Can we somehow make it easier/automate/validate?

For sure OSM tags are familiar to the people, however its flexibility brings a lot of complexity also, e.g. one feature could be tagged using different tag combinations. How do people know which combinations are supported by OM? Also both combinations could be used in styles definitions sometimes where they conflict each other... Whereas internal OM tags are in a straightforward hierarchy - much more simple.
As I mentioned before the biggest issue is a complex transformation between OSM tags and OM types. One needs to be aware of different tags combinations in `mapcss-mapping.csv` and `replacement_tags.txt`. Then there are some tags not present in OSM, but "invented" by OM, e.g. `[hwtag]`, `[city=]` (its a very ugly workaround - we got to invent non-existing tags for our special internal types to be able to convert them back to internal types!).
ATM I don't see a reasonable way to keep OSM tags in style definitions and make it simpler and more predictable at the same time...

**Comment by biodranik** [comment]
Ok. Then we need a validator that will complain on any incorrect non-OM tags in styles.

**Comment by endim8** [comment]
It'd be good if we could re-colour icons from drules, this would be useful for dark-style icons & cases like `phone` & `emergency-phone`, where emergency-phone is just a red version.

Probably requires vector icons though, unless we convert current icons to masks?

**Comment by pastk** [comment]
> It'd be good if we could re-colour icons from drules, this would be useful for dark-style icons & cases like `phone` & `emergency-phone`, where emergency-phone is just a red version.
> 
> Probably requires vector icons though, unless we convert current icons to masks?

Yeap the icons system could be greatly improved, see #2026

**Comment by pastk** [comment]
> One of the most important features we need from styles is dynamic map density. E.g. to show some objects earlier on a map if there is nothing else in the area. Should refactored styles system support it? Or should it be a code "hack"?
	
I see a few options which might fit different use cases.

E.g. for a "is the railway inside a city" use case we might want to make it thinner and visible at higher zoom levels. Or e.g. for an urban road we might want to show shields at later zooms, but for a country side road draw a contrast casing to make it more visible.
It involves changing the style (not just feature visibility).
We might use an existing "runtime conditions" mechanism for that (e.g. we use it to alter styling depending on city's population).
Actually in most (if not all) cases those conditions could be off-loaded from the runtime and resolved at the maps generation stage. The simplest solution could be to have e.g. `railway-rail` and `railway-rail-urban` types (or store an additional conditional attribute somehow else in feature's data).
	
If there will be too many types with such conditions it might clutter the styles though, so might be better to switch the whole style, e.g. from `main` to `main-urban` when inside city's boundaries.	
	
For simpler feature-level visibility changes only (i.e. when style definitions doesn't have to change) we might use an existing visibility index functionality.
E.g. if there are too many waterfalls in a particular cell of the index then just index the feature for this zoom level.

**Comment by biodranik** [comment]
There are many styles, but only one index. When we developed it, it was an ideal combination of 1 style and 1 index, so everything was optimized. Now we want to support many styles, and index in many cases will have most or many features in it.

Speaking about dynamic styles, I meant that if you're looking at the empty map it's not cool. If there are some objects on the map in that empty area that can be displayed without breaking something else, then it's better to display them (e.g. draw villages/lakes/rivers/forests earlier than they're specified now in styles, the same can be applied to POIs, isn't it cool to see something in a middle of a "desert"?).

**Comment by pastk** [comment]
> There are many styles, but only one index. When we developed it, it was an ideal combination of 1 style and 1 index, so everything was optimized. Now we want to support many styles, and index in many cases will have most or many features in it.
> 

True. But there are cases still that are not style-specific I think. E.g. a tight "web" of smaller streets in cities at z11-12.

> Speaking about dynamic styles, I meant that if you're looking at the empty map it's not cool. If there are some objects on the map in that empty area that can be displayed without breaking something else, then it's better to display them (e.g. draw villages/lakes/rivers/forests earlier than they're specified now in styles, the same can be applied to POIs, isn't it cool to see something in a middle of a "desert"?).

Yes. Ideally (when dynamic visibility is working fine) we'd have feature's visibility in styles not set according to its importance (like it is now, more or less), but just as a reasonable lower limit - e.g. we won't want a bench displayed at z12 still - even though it's nothing else around. Otherwise a feature's "importance" (and hence when to display it) should be derived from the priority and map "density" around at specific zoom.

**Comment by pastk** [comment]
What do we need to do to lower technical barriers for
1. contributing style designers
2. enthusiastic users wishing to use a custom style

?

The current process is too complex / too time consuming for the most people:
- clone the repo
- setup the build environment
- edit the style files
- run the generate_drules.sh script
- for mobile targets: re-build the app or push the compiled style file

Ideally, it should be possible to just edit the style files and see the changes in the app immediately.
Its possible, but its a lot of work - we'll get there some day.

Meanwhile are there any easier and still efficient solutions?
Can we skip the setup part?

Is it possible to wrap our python scripts into an easily installable cross-platform binary to compile the style files?
A desktop OM version should be also packaged into an easily installable package that includes the style compiler.
- download the style designer package
- edit the files
- run the style compiler (or press a button in the designer)
- push resulting files onto device to use the custom style or create a PR via github web ui to contribute

**Comment by biodranik** [comment]
Designer binary target was likely created for this purpose. Let's check how it worked.

**Comment by pastk** [comment]
Linking latest related PRs:
- https://github.com/organicmaps/organicmaps/pull/6019
- https://github.com/organicmaps/organicmaps/pull/6034
- https://github.com/organicmaps/organicmaps/pull/6015

**Comment by pastk** [comment]
Open discussions on related future ideas:
- https://github.com/organicmaps/organicmaps/issues/6331
- #6223
- #1312
